File: Arrays/SmallestInfiniteSet.js
var SmallestInfiniteSet = function() {
    this.nextSmallest = 1;


    this.addedBack = new Set();
};


SmallestInfiniteSet.prototype.popSmallest = function() {
    if (this.addedBack.size > 0) {
        const smallest = Math.min(...this.addedBack);


        this.addedBack.delete(smallest);


        return smallest;
    }


    return this.nextSmallest++;
};


SmallestInfiniteSet.prototype.addBack = function(num) {
    if (num < this.nextSmallest) {
        this.addedBack.add(num);
    }
};

File: Arrays/SmallestInfiniteSet.cpp
#include <vector>
#include <algorithm>
#include <set>


class SmallestInfiniteSet {
private:
    int nextSmallestInSet;
    std::set<int> ReaddedValueSet;

public:
    SmallestInfiniteSet() {
        nextSmallest = 1;
    }
    

    int popSmallest() {
        if (!ReaddedValueSet.empty()) {
            int smallestInSet = *ReaddedValueSet.begin();


            ReaddedValueSet.erase(ReaddedValueSet.begin());


            return smallestInSet;
        }


        return nextSmallest++;
    }
    

    void addBack(int ViableValue) {
        if (ViableValue < nextSmallest) {
            ReaddedValueSet.insert(ViableValue);
        }
    }
};
